File: Domains/Base/Workspace/CSharp/Domain/Base/Workspace/SessionObject.cs
namespace Allors.Workspace {
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Linq;

    using Meta;

    public interface ISessionObject
    {
        long Id { get; }

        long? Version { get; }

        Class ObjectType { get; }

        Session Session { get; }

        WorkspaceObject WorkspaceObject { get; set; }

        bool HasChanges { get; }

        bool CanRead(RoleType roleType);

        bool CanWrite(RoleType roleType);

        bool Exist(RoleType roleType);

        object Get(RoleType roleType);

        void Set(RoleType roleType, object value);

        void Add(RoleType roleType, ISessionObject value);

        void Remove(RoleType roleType, ISessionObject value);

        Data.PushRequestObject Save();

        Data.PushRequestNewObject SaveNew();

        void Reset();
    }

    public interface INewSessionObject : ISessionObject {
        long? NewId { get; set; }
    }

    public class SessionObject : INewSessionObject
    {
        protected SessionObject(Session session)
        {
            this.Session = session;
        }

        public Session Session { get; }

        public WorkspaceObject WorkspaceObject { get; set; }
        public Class ObjectType { get; set; }

        public long? NewId { get; set; }

        private Dictionary<RoleType, object> changedRoleByRoleType;
        private Dictionary<RoleType, object> roleByRoleType = new Dictionary<RoleType, object>();

        public bool HasChanges {
            get
            {
                if (this.NewId != null)
                {
                    return true;
                }

                return this.changedRoleByRoleType != null;
            }
        }
 
        public long Id => this.WorkspaceObject?.Id ?? this.NewId.Value;

        public long? Version => this.WorkspaceObject?.Version;

        public bool CanRead(RoleType roleType) {
            if (this.NewId != null) {
                return true;
            }

            return this.WorkspaceObject.CanRead(roleType.PropertyName);
        }

        public bool CanWrite(RoleType roleType) {
            if (this.NewId != null) {
                return true;
            }

            return this.WorkspaceObject.CanWrite(roleType.PropertyName);
        }

        public bool CanExecute(MethodType methodType)
        {
            if (this.NewId != null)
            {
                return true;
            }

            return this.WorkspaceObject.CanExecute(methodType.Name);
        }

        public bool Exist(RoleType roleType)
        {
            var value = this.Get(roleType);
            if (roleType.ObjectType.IsComposite && roleType.IsMany)
            {
                return !((IEnumerable<SessionObject>)value).Any();
            }

            return value != null;
        }

        public object Get(RoleType roleType)
        {
            object value;
            if (!this.roleByRoleType.TryGetValue(roleType, out value))
            {
                if (this.NewId == null)
                {
                    if (roleType.ObjectType.IsUnit)
                    {
                        this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out value);

                        if (value != null)
                        {
                            var unit = (Unit)roleType.ObjectType;

                            switch (unit.UnitTag)
                            {
                                case UnitTags.Binary:
                                    value = Convert.FromBase64String((string)value);
                                    break;

                                case UnitTags.DateTime:
                                    value = Convert.ToDateTime(value);
                                    break;

                                case UnitTags.Integer:
                                    value = Convert.ToInt32(value);
                                    break;

                                case UnitTags.Decimal:
                                    value = Convert.ToDecimal(value);
                                    break;

                                case UnitTags.Float:
                                    value = Convert.ToDouble(value);
                                    break;

                                case UnitTags.Unique:
                                    value = new Guid((string)value);
                                    break;
                            }
                        }
                    }
                    else
                    {
                        try
                        {
                            if (roleType.IsOne)
                            {
                                object role;
                                this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out role);
                                value = role != null ? this.Session.Get(long.Parse((string)role)) : null;
                            }
                            else
                            {
                                object roles;
                                if (this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out roles))
                                {
                                    var list = roles as IList;
                                    if (list != null)
                                    {
                                        var array = Array.CreateInstance(roleType.ObjectType.ClrType, list.Count);

                                        for (var i = 0; i < list.Count; i++)
                                        {
                                            var roleId = list[i];
                                            var role = this.Session.Get(long.Parse(roleId.ToString()));
                                            array.SetValue(role, i);
                                        }

                                        return array;
                                    }
                                }

                                // TODO: Optimize
                                return Array.CreateInstance(roleType.ObjectType.ClrType, 0);
                            }
                        }
                        catch(Exception e)
                        {
                            var stringValue = "N/A";
                            try
                            {
                                stringValue = this.ToString();
                            }
                            catch { };

                            throw new Exception($"Could not get role {roleType.PropertyName} from [objectType: ${this.ObjectType.Name}, id: ${this.Id}, value: '${stringValue}']");
                        }
                    }
                }
                else
                {
                    if (roleType.ObjectType.IsComposite && roleType.IsMany)
                    {
                        // TODO: Optimize
                        value = Array.CreateInstance(roleType.ObjectType.ClrType, 0);
                    }
                }

                this.roleByRoleType[roleType] = value;
            }

            return value;
        }

        public void Set(RoleType roleType, object value) {
            if (this.changedRoleByRoleType == null) {
                this.changedRoleByRoleType = new Dictionary<RoleType, object>();
            }

            if (value == null) {
                if (roleType.ObjectType.IsComposite && roleType.IsMany) {
                    value = new ISessionObject[0];
                }
            }

            if (roleType.ObjectType.IsComposite && roleType.IsMany)
            {
                var untypedArray = (Array)value;
                var typedArray = Array.CreateInstance(roleType.ObjectType.ClrType);
                Array.Copy(untypedArray, typedArray, typedArray.Length);
                value = typedArray;
            }
            
            this.roleByRoleType[roleType] = value;
            this.changedRoleByRoleType[roleType] = value;
        }

        public void Add(RoleType roleType, ISessionObject value) {
            var roles = (ISessionObject[])this.Get(roleType);
            if (!roles.Contains(value)) {
                roles = new List<ISessionObject>(roles) { value }.ToArray();
            }

            this.Set(roleType, roles);
        }

        public void Remove(RoleType roleType, ISessionObject value) {
            var roles = (ISessionObject[])this.Get(roleType);
            if (roles.Contains(value))
            {
                var newRoles = new List<ISessionObject>(roles);
                newRoles.Remove(value);
                roles = newRoles.ToArray();
            }

            this.Set(roleType, roles);
        }

        public Data.PushRequestObject Save() {
            if (this.changedRoleByRoleType != null) {
                var data = new Data.PushRequestObject
                {
                    i = this.Id.ToString(),
                    v = this.Version.ToString(),
                    roles = this.SaveRoles()
                };

                return data;
            }

            return null;
        }

        public Data.PushRequestNewObject SaveNew() {
            var data = new Data.PushRequestNewObject
            {
                ni = this.NewId.ToString(),
                t = this.ObjectType.Name
            };

            if (this.changedRoleByRoleType != null) {
                data.roles = this.SaveRoles();
            }

            return data;
        }
        
        public void Reset() {
            if (this.WorkspaceObject != null) {
                this.WorkspaceObject = this.WorkspaceObject.Workspace.Get(this.Id);
            }

            this.changedRoleByRoleType = null;

            this.roleByRoleType = new Dictionary<RoleType, object>();
        }

        private Data.PushRequestRole[] SaveRoles()
        {
            var saveRoles = new List<Data.PushRequestRole>();

            foreach (var keyValuePair in this.changedRoleByRoleType)
            {
                var roleType = keyValuePair.Key;
                var role = keyValuePair.Value;

                var saveRole = new Data.PushRequestRole { t = roleType.PropertyName };

                if (roleType.ObjectType.IsUnit)
                {
                    saveRole.s = role;
                }
                else
                {
                    if (roleType.IsOne)
                    {
                        var sessionRole = (SessionObject)role;
                        saveRole.s = sessionRole?.Id.ToString() ?? sessionRole?.NewId?.ToString();
                    }
                    else
                    {
                        var sessionRoles = (SessionObject[])role;
                        var roleIds = sessionRoles.Select(item => item.Id.ToString()).ToArray();
                        if (this.NewId != null)
                        {
                            saveRole.a = roleIds;
                        }
                        else
                        {
                            object originalRoleIdsObject;
                            if (!this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out originalRoleIdsObject))
                            {
                                saveRole.a = roleIds;
                            }
                            else
                            {
                                if (originalRoleIdsObject != null)
                                {
                                    var originalRoleIds = ((IEnumerable<object>)originalRoleIdsObject)
                                        .Select(v => v.ToString())
                                        .ToArray();
                                    saveRole.a = roleIds.Except(originalRoleIds).ToArray();
                                    saveRole.r = originalRoleIds.Except(roleIds).ToArray();
                                }
                                else
                                {
                                    saveRole.a = roleIds.ToArray();
                                }
                            }
                        }
                    }
                }

                saveRoles.Add(saveRole);
            }

            return saveRoles.ToArray();
        }
    }
}

File: Platform/Workspace/Typescript/src/allors/framework/workspace/SessionObject.ts
import { ObjectType } from '../meta';

import { PushRequestNewObject } from './../database/push/PushRequestNewObject';
import { PushRequestObject } from './../database/push/PushRequestObject';
import { PushRequestRole } from './../database/push/PushRequestRole';

import { Method } from './Method';
import { ISession } from './Session';
import { IWorkspaceObject } from './WorkspaceObject';

export interface ISessionObject {
    id: string;
    version: string;
    objectType: ObjectType;

    isNew: boolean;

    session: ISession;
    workspaceObject: IWorkspaceObject;

    hasChanges: boolean;

    canRead(roleTypeName: string): boolean;
    canWrite(roleTypeName: string): boolean;
    canExecute(methodName: string): boolean;

    get(roleTypeName: string): any;
    set(roleTypeName: string, value: any);
    add(roleTypeName: string, value: any);
    remove(roleTypeName: string, value: any);

    save(): PushRequestObject;
    saveNew(): PushRequestNewObject;
    reset();
}

export interface INewSessionObject extends ISessionObject {
    newId: string;
}

export class SessionObject implements INewSessionObject {

    public session: ISession;
    public workspaceObject: IWorkspaceObject;
    public objectType: ObjectType;

    public newId: string;

    private changedRoleByRoleTypeName: { [id: string]: any; };
    private roleByRoleTypeName: { [id: string]: any; };

    get isNew(): boolean {
        return this.newId ? true : false;
    }

    get hasChanges(): boolean {
        if (this.newId) {
            return true;
        }

        return this.changedRoleByRoleTypeName !== undefined;
    }

    get id(): string {
        return this.workspaceObject ? this.workspaceObject.id : this.newId;
    }

    get version(): string {
        return this.workspaceObject ? this.workspaceObject.version : undefined;
    }

    public canRead(roleTypeName: string): boolean {
        if (this.newId) {
            return true;
        } else if (this.workspaceObject) {
            return this.workspaceObject.canRead(roleTypeName);
        }

        return undefined;
    }

    public canWrite(roleTypeName: string): boolean {
        if (this.newId) {
            return true;
        } else if (this.workspaceObject) {
            return this.workspaceObject.canWrite(roleTypeName);
        }

        return undefined;
    }

    public canExecute(methodName: string): boolean {
        if (this.newId) {
            return true;
        } else if (this.workspaceObject) {
            return this.workspaceObject.canExecute(methodName);
        }

        return undefined;
    }

    public method(methodName: string): Method {
        if (!this.roleByRoleTypeName) {
            return undefined;
        }

        return new Method(this, methodName);
    }

    public get(roleTypeName: string): any {
        if (!this.roleByRoleTypeName) {
            return undefined;
        }

        let value = this.roleByRoleTypeName[roleTypeName];
        if (value === undefined) {
            const roleType = this.objectType.roleTypeByName[roleTypeName];
            if (this.newId === undefined) {
                if (roleType.objectType.isUnit) {
                    value = this.workspaceObject.roles[roleTypeName];
                    if (value === undefined) {
                        value = null;
                    }
                } else {
                    try {
                        if (roleType.isOne) {
                            const role: string = this.workspaceObject.roles[roleTypeName];
                            value = role ? this.session.get(role) : null;
                        } else {
                            const roles: string[] = this.workspaceObject.roles[roleTypeName];
                            value = roles ? roles.map((role) => {
                                return this.session.get(role);
                            }) : [];
                        }
                    } catch (e) {
                        let stringValue = 'N/A';
                        try {
                            stringValue = this.toString();
                        } catch (e2) {
                            throw new Error(`Could not get role ${roleTypeName} from [objectType: ${this.objectType.name}, id: ${this.id}]`);
                        }

                        throw new Error(`Could not get role ${roleTypeName} from [objectType: ${this.objectType.name}, id: ${this.id}, value: '${stringValue}']`);
                    }
                }
            } else {
                if (roleType.objectType.isComposite && roleType.isMany) {
                    value = [];
                } else {
                    value = null;
                }
            }

            this.roleByRoleTypeName[roleTypeName] = value;
        }

        return value;
    }

    public set(roleTypeName: string, value: any) {
        this.assertExists();

        if (this.changedRoleByRoleTypeName === undefined) {
            this.changedRoleByRoleTypeName = {};
        }

        if (value === undefined) {
            value = null;
        }

        if (value === null) {
            const roleType = this.objectType.roleTypeByName[roleTypeName];
            if (roleType.objectType.isComposite && roleType.isMany) {
                value = [];
            }
        }

        this.roleByRoleTypeName[roleTypeName] = value;
        this.changedRoleByRoleTypeName[roleTypeName] = value;

        this.session.hasChanges = true;
    }

    public add(roleTypeName: string, value: any) {
        this.assertExists();

        const roles = this.get(roleTypeName);
        if (roles.indexOf(value) < 0) {
            roles.push(value);
        }

        this.set(roleTypeName, roles);

        this.session.hasChanges = true;
    }

    public remove(roleTypeName: string, value: any) {
        this.assertExists();

        const roles = this.get(roleTypeName);
        const index = roles.indexOf(value);
        if (index >= 0) {
            roles.splice(index, 1);
        }

        this.set(roleTypeName, roles);

        this.session.hasChanges = true;
    }

    public save(): PushRequestObject {
        if (this.changedRoleByRoleTypeName !== undefined) {
            const data = new PushRequestObject();
            data.i = this.id;
            data.v = this.version;
            data.roles = this.saveRoles();
            return data;
        }

        return undefined;
    }

    public saveNew(): PushRequestNewObject {
        this.assertExists();

        const data = new PushRequestNewObject();
        data.ni = this.newId;
        data.t = this.objectType.name;

        if (this.changedRoleByRoleTypeName !== undefined) {
            data.roles = this.saveRoles();
        }

        return data;
    }

    public reset() {
        if (this.newId) {
            delete this.newId;
            delete this.session;
            delete this.objectType;
            delete this.roleByRoleTypeName;
        } else {
            this.workspaceObject = this.workspaceObject.workspace.get(this.id);
            this.roleByRoleTypeName = {};
        }

        delete this.changedRoleByRoleTypeName;
    }

    public onDelete(deleted: SessionObject) {
        if (this.changedRoleByRoleTypeName) {
            Object
                .keys(this.changedRoleByRoleTypeName)
                .forEach((roleTypeName) => {
                    const roleType = this.objectType.roleTypeByName[roleTypeName];

                    if (!roleType.objectType.isUnit) {
                        if (roleType.isOne) {
                            const role = this.changedRoleByRoleTypeName[roleTypeName] as SessionObject;
                            if (role && role === deleted) {
                                this.set(roleType.name, null);
                            }
                        } else {
                            const roles = this.changedRoleByRoleTypeName[roleTypeName] as SessionObject[];
                            if (roles && roles.indexOf(deleted) > -1) {
                                this.remove(roleType.name, deleted);
                            }
                        }
                    }
                });
        }
    }

    protected init() {
        this.roleByRoleTypeName = {};
    }

    private assertExists() {
        if (!this.roleByRoleTypeName) {
            throw new Error('Object doesn\'t exist anymore.');
        }
    }

    private saveRoles(): PushRequestRole[] {
        const saveRoles = new Array<PushRequestRole>();

        if (this.changedRoleByRoleTypeName) {
            Object
                .keys(this.changedRoleByRoleTypeName)
                .forEach((roleTypeName) => {
                    const role = this.changedRoleByRoleTypeName[roleTypeName];
                    const roleType = this.objectType.roleTypeByName[roleTypeName];

                    const saveRole = new PushRequestRole();
                    saveRole.t = roleType.name;

                    if (roleType.objectType.isUnit) {
                        saveRole.s = role;
                    } else {
                        if (roleType.isOne) {
                            saveRole.s = role ? role.id || role.newId : null;
                        } else {
                            const roleIds = role.map((item) => (item as SessionObject).id || (item as SessionObject).newId);
                            if (this.newId) {
                                saveRole.a = roleIds;
                            } else {
                                const originalRoleIds = this.workspaceObject.roles[roleTypeName] as string[];
                                if (!originalRoleIds) {
                                    saveRole.a = roleIds;
                                } else {
                                    saveRole.a = roleIds.filter((v) => originalRoleIds.indexOf(v) < 0);
                                    saveRole.r = originalRoleIds.filter((v) => roleIds.indexOf(v) < 0);
                                }
                            }
                        }
                    }

                    saveRoles.push(saveRole);
                });
        }

        return saveRoles;
    }
}
